File: exams/travez/u3/Exam 3/espe_edu_ec_sorter_controller/Sorting_Context.py
from abc import ABC, abstractmethod
from typing import List

from Bubble_Sort import Bubble_Sort


class SortingContext:
    
    def __init__(self):
        self._strategy = None
    
    def _select_strategy(self, data: List[int]):
        size = len(data)
        if 2 <= size <= 5:
            self._strategy = BubbleSort()
        elif 6 <= size <= 10:
            self._strategy = InsertionSort()
        else:
            self._strategy = QuickSort()
    
    def sort_data(self, data: List[int]) -> List[int]:
        self._select_strategy(data)
        return self._strategy.sort(data)

File: hw/pantoja/u3/HW29TemplateAndCompsiteandStrategy/HW30Strategy/src/main/java/espe/edu/ec/SortApp/controller/SortingContext.java
package espe.edu.ec.SortApp.controller;

import espe.edu.ec.SortApp.model.BubbleSort;
import espe.edu.ec.SortApp.model.InsertionSort;
import espe.edu.ec.SortApp.model.QuickSort;

/**
 *
 * @author David Pantoja, JavaSquad, DCCO-ESPE
 */

public class SortingContext {
    private SortingStrategy ss;

    public int[] sort(int data[]) {
        int size = data.length;
        ss = setSortStrategy(size);
        return ss.sort(data);
    }

    // Method to choose the sorting strategy based on the number of elements
    public SortingStrategy setSortStrategy(int n) {
        if (n > 0 && n < 30) {
            ss = new BubbleSort();  // Use BubbleSort for 1 to 29 elements
        } else if (n >= 30 && n < 100) {
            ss = new InsertionSort();  // Use InsertionSort for 30 to 99 elements
        } else if (n >= 100) {
            ss = new QuickSort();  // Use QuickSort for 100 or more elements
        }
        return ss;
    }
}